contrib/src/main/java/org/apache/apex/malhar/contrib/kudu/AbstractKuduOutputOperator.java

```diff
@@ -0,0 +1,666 @@
+ * properties to Kudu. The connection properties are set using the {@link ApexKuduConnection} using a builder pattern.
+ * </li>
+ * <li>Implement the logic how tuples need to ne handled in the event of a reconciliation phase ( i.e. when an
+ * operator is resuming back from failure ). See javadoc of the method for more details.</li>
+ * <li>Define the payload class</li>
+ * </ol>
+ * </p>
+ * <p>
+ * Note that the tuple that is getting processed is not the POJO class. The tuple that is getting processed is the
+ * {@link KuduExecutionContext} instance. This is because the operator supports mutation types as a higher level
+ * construct than simply writing a POJO into a Kudu table row.
+ * </p>
+ * <p>
+ * Supported mutations are:
+ * <ol>
+ *      <li>INSERT</li>
+ *      <li>UPDATE</li>
+ *      <li>UPSERT</li>
+ *      <li>DELETE</li>
+ * </ol>
+ * </p>
+ * <p>
+ * Please note that the Update mutation allows to change a subset of columns only even though there might be columns
+ * that were defined to be non-nullable. This is because the original mutation of type insert would have written the
+ * mandatory columns. In such scenarios, the method setDoNotWriteColumns() in {@link KuduExecutionContext} can be
+ * used to specify only those columns that need an update. This ways a read and Update pattern can be merged to a
+ * simple update pattern thus avoiding a read if required.</p>
+ * */
+@InterfaceStability.Evolving
+public abstract class AbstractKuduOutputOperator extends BaseOperator
+    implements Operator.ActivationListener<Context.OperatorContext>,Operator.CheckpointNotificationListener
+{
+
+  private ApexKuduConnection apexKuduConnection;
+
+  private transient  KuduTable kuduTable;
+
+  private transient  KuduSession kuduSession;
+
+  private transient KuduClient kuduClientHandle;
+
+  private transient Map<String,ColumnSchema> allColumnDefs;
+
+  private transient Map<String,Object> kuduColumnBasedGetters;
+
+  private Set<String> primaryKeyColumnNames;
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(AbstractKuduOutputOperator.class);
+
+  private WindowDataManager windowDataManager;
+
+  private transient long currentWindowId;
+
+  private transient boolean isInSafeMode;
+
+  private transient boolean isInReconcilingMode;
+
+  private transient long reconcilingWindowId;
+
+  @AutoMetric
+  transient long numInsertsInThisWindow;
+
+  @AutoMetric
+  transient long numUpsertsInThisWindow;
+
+  @AutoMetric
+  transient long numDeletedInThisWindow;
+
+  @AutoMetric
+  transient long numUpdatesInThisWindow;
+
+  @AutoMetric
+  transient long opsErrorsInThisWindow;
+
+  @AutoMetric
+  transient long bytesWrittenInThisWindow;
+
+  @AutoMetric
+  transient long rpcErrorsInThisWindow;
+
+  @AutoMetric
+  transient long writeOpsInThisWindow;
+
+  @AutoMetric
+  transient long writeRPCsInThisWindow;
+
+  @AutoMetric
+  long totalOpsErrors = 0;
+
+  @AutoMetric
+  long totalBytesWritten = 0;
```

Comment:
This metric is a different one. The totals of course are simply summed up but that is only applicable since the application start ( and those metrics are transient by nature). The ones above are named total because they represent the total metrics since the applications first launch ( and hence not marked transient for the same reason )